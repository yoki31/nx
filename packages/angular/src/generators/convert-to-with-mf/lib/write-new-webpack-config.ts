import type { SourceFile } from 'typescript';
import {
  getExposedModulesFromRemote,
  getRemotesFromHost,
  IsHostRemoteConfigResult,
} from './is-host-remote-config';

export function writeNewWebpackConfig(
  ast: SourceFile,
  mfType: IsHostRemoteConfigResult,
  projectName: string
) {
  const webpackConfig = `const { withModuleFederation } = require('@nx/angular/module-federation');
const config = require('./module-federation.config');
module.exports = withModuleFederation(config);
`;

  let mfeConfig = '';
  if (!mfType) {
    mfeConfig = `
        module.exports = {
          name: '${projectName}',
        };`;
  } else if (mfType === 'host') {
    const remotes = hostRemotesToString(ast);
    mfeConfig = `
        module.exports = {
          name: '${projectName}',
          remotes: ${remotes},
        };`;
  } else if (mfType === 'remote') {
    const exposedModules = getExposedModulesFromRemote(ast);
    mfeConfig = `
        module.exports = {
          name: '${projectName}',
          exposes: ${exposedModules},
        };`;
  } else if (mfType === 'both') {
    const remotes = hostRemotesToString(ast);
    const exposedModules = getExposedModulesFromRemote(ast);
    mfeConfig = `
    module.exports = {
      name: '${projectName}',
      remotes: ${remotes},
      exposes: ${exposedModules},
    };`;
  }

  return [webpackConfig, mfeConfig];
}

function hostRemotesToString(ast: SourceFile) {
  const remotes: string = getRemotesFromHost(ast)
    .reduce(
      (acc, remotePair) => `['${remotePair[0]}', '${remotePair[1]}'], ${acc}`,
      ''
    )
    .trim();
  return `[${remotes.endsWith(',') ? remotes.slice(0, -1) : remotes}]`;
}
