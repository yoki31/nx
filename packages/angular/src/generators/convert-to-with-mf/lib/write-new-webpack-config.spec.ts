import { tsquery } from '@phenomnomnominal/tsquery';
import { writeNewWebpackConfig } from './write-new-webpack-config';

describe('writeNewWebpackConfig', () => {
  it('should convert host config correctly', () => {
    // ARRANGE
    const sourceText = `module.exports = {
        plugins: [
            new ModuleFederationPlugin({
                remotes: {
                    remote1: 'http://localhost:4201/remoteEntry.mjs'
                }
            })
        ]
    }`;

    const ast = tsquery.ast(sourceText);

    // ACT
    const result = writeNewWebpackConfig(ast, 'host', 'host1');

    // ASSERT
    expect(result).toMatchSnapshot();
  });

  it('should convert remote config correctly', () => {
    // ARRANGE
    const sourceText = `module.exports = {
        plugins: [
            new ModuleFederationPlugin({
                exposes: {
                    './Module': 'apps/remote1/src/app/remote-entry/entry.module.ts'
                }
            })
        ]
    }`;

    const ast = tsquery.ast(sourceText);

    // ACT
    const result = writeNewWebpackConfig(ast, 'remote', 'remote1');

    // ASSERT
    expect(result).toMatchSnapshot();
  });

  it('should convert config that is both remote and host correctly', () => {
    // ARRANGE
    const sourceText = `module.exports = {
        plugins: [
            new ModuleFederationPlugin({
                remotes: {
                    remote1: 'http://localhost:4201/remoteEntry.mjs'
                },
                exposes: {
                    './Module': 'apps/both/src/app/remote-entry/entry.module.ts'
                }
            })
        ]
    }`;

    const ast = tsquery.ast(sourceText);

    // ACT
    const result = writeNewWebpackConfig(ast, 'both', 'both1');

    // ASSERT
    expect(result).toMatchSnapshot();
  });

  it('should convert config that is neither remote and host correctly', () => {
    // ARRANGE
    const sourceText = `module.exports = {
        plugins: [
            new ModuleFederationPlugin({})
        ]
    }`;

    const ast = tsquery.ast(sourceText);

    // ACT
    const result = writeNewWebpackConfig(ast, false, 'neither');

    // ASSERT
    expect(result).toMatchSnapshot();
  });
});
