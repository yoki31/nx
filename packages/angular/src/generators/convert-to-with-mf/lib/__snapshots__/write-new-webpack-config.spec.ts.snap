// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`writeNewWebpackConfig should convert config that is both remote and host correctly 1`] = `
[
  "const { withModuleFederation } = require('@nx/angular/module-federation');
const config = require('./module-federation.config');
module.exports = withModuleFederation(config);
",
  "
    module.exports = {
      name: 'both1',
      remotes: [['remote1', 'http://localhost:4201']],
      exposes: {
                    './Module': 'apps/both/src/app/remote-entry/entry.module.ts'
                },
    };",
]
`;

exports[`writeNewWebpackConfig should convert config that is neither remote and host correctly 1`] = `
[
  "const { withModuleFederation } = require('@nx/angular/module-federation');
const config = require('./module-federation.config');
module.exports = withModuleFederation(config);
",
  "
        module.exports = {
          name: 'neither',
        };",
]
`;

exports[`writeNewWebpackConfig should convert host config correctly 1`] = `
[
  "const { withModuleFederation } = require('@nx/angular/module-federation');
const config = require('./module-federation.config');
module.exports = withModuleFederation(config);
",
  "
        module.exports = {
          name: 'host1',
          remotes: [['remote1', 'http://localhost:4201']],
        };",
]
`;

exports[`writeNewWebpackConfig should convert remote config correctly 1`] = `
[
  "const { withModuleFederation } = require('@nx/angular/module-federation');
const config = require('./module-federation.config');
module.exports = withModuleFederation(config);
",
  "
        module.exports = {
          name: 'remote1',
          exposes: {
                    './Module': 'apps/remote1/src/app/remote-entry/entry.module.ts'
                },
        };",
]
`;
